File: schedule.go
// schedule.go
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

type Connection struct {
	From      FromTo   `json:"from"`
	To        FromTo   `json:"to"`
	Duration  string   `json:"duration"`
	Transfers int      `json:"transfers"`
}

type FromTo struct {
	Departure string `json:"departure"`
	Arrival   string `json:"arrival"`
	Station   struct {
		Name string `json:"name"`
	} `json:"station"`
}

type ConnectionsResponse struct {
	Connections []Connection `json:"connections"`
}

var cache = make(map[string]struct {
	Data      []Connection
	Timestamp time.Time
})
const cacheTTL = 60 * time.Second

func fetchConnections(from, to, date, timeStr string) ([]Connection, error) {
	key := fmt.Sprintf("%s:%s:%s:%s", from, to, date, timeStr)
	if entry, ok := cache[key]; ok {
		if time.Since(entry.Timestamp) < cacheTTL {
			return entry.Data, nil
		}
	}
	apiURL := "https://transport.opendata.ch/v1/connections"
	params := url.Values{}
	params.Set("from", from)
	params.Set("to", to)
	if date != "" {
		params.Set("date", date)
	}
	if timeStr != "" {
		params.Set("time", timeStr)
	}
	fullURL := apiURL + "?" + params.Encode()
	resp, err := http.Get(fullURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data ConnectionsResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	cache[key] = struct {
		Data      []Connection
		Timestamp time.Time
	}{Data: data.Connections, Timestamp: time.Now()}
	return data.Connections, nil
}

func display(conns []Connection, limit int, sortBy string) {
	if len(conns) == 0 {
		fmt.Println("No connections found.")
		return
	}
	// format and sort
	type displayConn struct {
		DepTime      string
		DepStation   string
		ArrTime      string
		ArrStation   string
		Duration     string
		Changes      int
	}
	var list []displayConn
	for _, c := range conns {
		depTime := ""
		if len(c.From.Departure) > 5 {
			depTime = c.From.Departure[11:16]
		}
		arrTime := ""
		if len(c.To.Arrival) > 5 {
			arrTime = c.To.Arrival[11:16]
		}
		list = append(list, displayConn{
			DepTime:    depTime,
			DepStation: c.From.Station.Name,
			ArrTime:    arrTime,
			ArrStation: c.To.Station.Name,
			Duration:   c.Duration,
			Changes:    c.Transfers,
		})
	}
	switch sortBy {
	case "departure":
		sort.Slice(list, func(i, j int) bool { return list[i].DepTime < list[j].DepTime })
	case "duration":
		sort.Slice(list, func(i, j int) bool { return list[i].Duration < list[j].Duration })
	case "changes":
		sort.Slice(list, func(i, j int) bool { return list[i].Changes < list[j].Changes })
	default:
		sort.Slice(list, func(i, j int) bool { return list[i].DepTime < list[j].DepTime })
	}
	fmt.Printf("\nFound %d connections.\n", len(list))
	fmt.Printf("%-3s %-20s %-20s %-10s %-8s\n", "#", "Departure", "Arrival", "Duration", "Changes")
	for i, c := range list {
		if i >= limit {
			fmt.Printf("... and %d more\n", len(list)-limit)
			break
		}
		depStr := fmt.Sprintf("%s %s", c.DepTime, truncate(c.DepStation, 10))
		arrStr := fmt.Sprintf("%s %s", c.ArrTime, truncate(c.ArrStation, 10))
		fmt.Printf("%-3d %-20s %-20s %-10s %-8d\n", i+1, depStr, arrStr, c.Duration, c.Changes)
	}
}

func truncate(s string, max int) string {
	if len(s) > max {
		return s[:max]
	}
	return s
}

func exportJSON(conns []Connection, filename string) {
	data, _ := json.MarshalIndent(conns, "", "  ")
	ioutil.WriteFile(filename, data, 0644)
	fmt.Printf("Exported to %s\n", filename)
}

func exportCSV(conns []Connection, filename string) {
	file, _ := os.Create(filename)
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"Departure", "Departure Station", "Arrival", "Arrival Station", "Duration", "Changes"})
	for _, c := range conns {
		writer.Write([]string{
			c.From.Departure,
			c.From.Station.Name,
			c.To.Arrival,
			c.To.Station.Name,
			c.Duration,
			strconv.Itoa(c.Transfers),
		})
	}
	fmt.Printf("Exported to %s\n", filename)
}

func interactive() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("=== Train Schedule ===")
	fmt.Print("From station: ")
	from, _ := reader.ReadString('\n')
	from = strings.TrimSpace(from)
	fmt.Print("To station: ")
	to, _ := reader.ReadString('\n')
	to = strings.TrimSpace(to)
	if from == "" || to == "" {
		fmt.Println("Stations required.")
		return
	}
	fmt.Print("Date (YYYY-MM-DD, leave blank for today): ")
	date, _ := reader.ReadString('\n')
	date = strings.TrimSpace(date)
	fmt.Print("Time (HH:MM, leave blank for now): ")
	timeStr, _ := reader.ReadString('\n')
	timeStr = strings.TrimSpace(timeStr)
	conns, err := fetchConnections(from, to, date, timeStr)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	display(conns, 10, "departure")
	if len(conns) > 0 {
		fmt.Print("\nExport results? (y/n): ")
		export, _ := reader.ReadString('\n')
		export = strings.TrimSpace(strings.ToLower(export))
		if export == "y" {
			fmt.Print("Format (json/csv): ")
			fmtF, _ := reader.ReadString('\n')
			fmtF = strings.TrimSpace(strings.ToLower(fmtF))
			fmt.Print("Filename (default: schedule." + fmtF + "): ")
			fname, _ := reader.ReadString('\n')
			fname = strings.TrimSpace(fname)
			if fname == "" {
				fname = "schedule." + fmtF
			}
			if fmtF == "json" {
				exportJSON(conns, fname)
			} else if fmtF == "csv" {
				exportCSV(conns, fname)
			} else {
				fmt.Println("Unknown format.")
			}
		}
	}
}

func cli() {
	from := flag.String("from", "", "Departure station")
	to := flag.String("to", "", "Destination station")
	date := flag.String("date", "", "Departure date (YYYY-MM-DD)")
	timeStr := flag.String("time", "", "Departure time (HH:MM)")
	limit := flag.Int("limit", 10, "Max connections")
	sortBy := flag.String("sort", "departure", "Sort field: departure, duration, changes")
	exportFmt := flag.String("export", "", "Export format: json or csv")
	output := flag.String("output", "", "Export filename")
	flag.Parse()
	if *from == "" || *to == "" {
		fmt.Println("--from and --to are required.")
		flag.Usage()
		return
	}
	conns, err := fetchConnections(*from, *to, *date, *timeStr)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	display(conns, *limit, *sortBy)
	if *exportFmt != "" && len(conns) > 0 {
		fname := *output
		if fname == "" {
			fname = "schedule." + *exportFmt
		}
		if *exportFmt == "json" {
			exportJSON(conns, fname)
		} else if *exportFmt == "csv" {
			exportCSV(conns, fname)
		}
	}
}

func main() {
	if len(os.Args) > 1 {
		cli()
	} else {
		interactive()
	}
}

File: schedule.swift
// schedule.swift
import Foundation

struct Connection: Codable {
    let from: FromTo
    let to: FromTo
    let duration: String
    let transfers: Int
}

struct FromTo: Codable {
    let departure: String
    let arrival: String
    let station: Station
}

struct Station: Codable {
    let name: String
}

struct Response: Codable {
    let connections: [Connection]?
}

var cache: [String: (data: [Connection], timestamp: TimeInterval)] = [:]
let cacheTTL: TimeInterval = 60

func fetchConnections(from: String, to: String, date: String?, time: String?) -> [Connection] {
    let key = "\(from):\(to):\(date ?? ""):\(time ?? "")"
    if let entry = cache[key] {
        if Date().timeIntervalSince1970 - entry.timestamp < cacheTTL {
            return entry.data
        }
    }
    var components = URLComponents(string: "https://transport.opendata.ch/v1/connections")!
    var queryItems = [URLQueryItem(name: "from", value: from), URLQueryItem(name: "to", value: to)]
    if let d = date { queryItems.append(URLQueryItem(name: "date", value: d)) }
    if let t = time { queryItems.append(URLQueryItem(name: "time", value: t)) }
    components.queryItems = queryItems
    guard let url = components.url else { return [] }
    let semaphore = DispatchSemaphore(value: 0)
    var result: [Connection] = []
    var error: Error? = nil
    let task = URLSession.shared.dataTask(with: url) { data, _, err in
        defer { semaphore.signal() }
        if let err = err { error = err; return }
        guard let data = data else { return }
        do {
            let decoder = JSONDecoder()
            let response = try decoder.decode(Response.self, from: data)
            result = response.connections ?? []
            cache[key] = (result, Date().timeIntervalSince1970)
        } catch {
            error = error
        }
    }
    task.resume()
    semaphore.wait()
    if let err = error {
        print("Error: \(err)")
        return []
    }
    return result
}

func display(connections: [Connection], limit: Int = 10, sortBy: String = "departure") {
    if connections.isEmpty {
        print("No connections found.")
        return
    }
    var sorted = connections
    switch sortBy {
    case "duration":
        sorted.sort { $0.duration < $1.duration }
    case "changes":
        sorted.sort { $0.transfers < $1.transfers }
    default:
        sorted.sort { $0.from.departure < $1.from.departure }
    }
    print("\nFound \(sorted.count) connections.")
    print(String(repeating: "-", count: 80))
    print(String(format: "%-3s %-20s %-20s %-10s %-8s", "#", "Departure", "Arrival", "Duration", "Changes"))
    for (i, c) in sorted.enumerated() {
        if i >= limit {
            print("... and \(sorted.count - limit) more")
            break
        }
        let depTime = c.from.departure.count > 5 ? String(c.from.departure.suffix(5)) : "?"
        let arrTime = c.to.arrival.count > 5 ? String(c.to.arrival.suffix(5)) : "?"
        let depSt = c.from.station.name.prefix(10)
        let arrSt = c.to.station.name.prefix(10)
        print(String(format: "%-3d %@ %@ %@ %@ %@ %-8d", i+1, depTime, depSt, arrTime, arrSt, c.duration, c.transfers))
    }
}

func exportJSON(connections: [Connection], filename: String) {
    let encoder = JSONEncoder()
    encoder.outputFormatting = .prettyPrinted
    do {
        let data = try encoder.encode(connections)
        try data.write(to: URL(fileURLWithPath: filename))
        print("Exported to \(filename)")
    } catch {
        print("Export failed: \(error)")
    }
}

func exportCSV(connections: [Connection], filename: String) {
    var lines = ["Departure,Departure Station,Arrival,Arrival Station,Duration,Changes"]
    for c in connections {
        lines.append("\(c.from.departure),\"\(c.from.station.name)\",\(c.to.arrival),\"\(c.to.station.name)\",\(c.duration),\(c.transfers)")
    }
    do {
        try lines.joined(separator: "\n").write(toFile: filename, atomically: true, encoding: .utf8)
        print("Exported to \(filename)")
    } catch {
        print("Export failed: \(error)")
    }
}

func interactive() {
    print("=== Train Schedule ===")
    print("From station: ", terminator: "")
    guard let from = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines), !from.isEmpty else {
        print("Station required.")
        return
    }
    print("To station: ", terminator: "")
    guard let to = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines), !to.isEmpty else {
        print("Station required.")
        return
    }
    print("Date (YYYY-MM-DD, leave blank for today): ", terminator: "")
    var date = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines)
    if date?.isEmpty ?? true { date = nil }
    print("Time (HH:MM, leave blank for now): ", terminator: "")
    var time = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines)
    if time?.isEmpty ?? true { time = nil }
    let conns = fetchConnections(from: from, to: to, date: date, time: time)
    display(connections: conns)
    if !conns.isEmpty {
        print("\nExport results? (y/n): ", terminator: "")
        guard let exportAns = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() else { return }
        if exportAns == "y" {
            print("Format (json/csv): ", terminator: "")
            guard let fmt = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() else { return }
            print("Filename (default: schedule.\(fmt)): ", terminator: "")
            var fname = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines)
            if fname?.isEmpty ?? true { fname = "schedule.\(fmt)" }
            if fmt == "json" {
                exportJSON(connections: conns, filename: fname!)
            } else if fmt == "csv" {
                exportCSV(connections: conns, filename: fname!)
            } else {
                print("Unknown format.")
            }
        }
    }
}

func cli() {
    let args = CommandLine.arguments.dropFirst()
    var params: [String: String] = [:]
    var i = 0
    while i < args.count {
        let arg = args[i]
        if arg.hasPrefix("--") {
            let key = String(arg.dropFirst(2))
            if i + 1 < args.count && !args[i+1].hasPrefix("--") {
                params[key] = args[i+1]
                i += 2
            } else {
                params[key] = "true"
                i += 1
            }
        } else {
            i += 1
        }
    }
    guard let from = params["from"], let to = params["to"] else {
        print("--from and --to are required.")
        return
    }
    let date = params["date"]
    let time = params["time"]
    let limit = Int(params["limit"] ?? "10") ?? 10
    let sortBy = params["sort"] ?? "departure"
    let exportFmt = params["export"]
    let output = params["output"]
    let conns = fetchConnections(from: from, to: to, date: date, time: time)
    display(connections: conns, limit: limit, sortBy: sortBy)
    if let expFmt = exportFmt, !conns.isEmpty {
        let fname = output ?? "schedule.\(expFmt)"
        if expFmt == "json" {
            exportJSON(connections: conns, filename: fname)
        } else if expFmt == "csv" {
            exportCSV(connections: conns, filename: fname)
        }
    }
}

if CommandLine.arguments.count > 1 {
    cli()
} else {
    interactive()
}
